### DISCUSSION TITLE
Update Document: Append to/Remove Array/Object

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#522] [Feedback & Feature Proposal]
https://github.com/meilisearch/meilisearch/discussions/2693
https://github.com/meilisearch/meilisearch/discussions/2694

Created by @geoffatsource

"Is it possible to append to an existing array or object in a document? And relatedly, is there a way to remove?

My particular use-case is that I want to be able to "tag" documents without having to know the existing tags already assigned to those documents."

---

"Is there a way to completely delete a field in an existing document without replacing the entire document?"

**Comment by intoxicated** [comment]
So it seems like theres no way to append or remove items from array, then it has to be updated with full list on every update?

**Reply by gmourier** [reply]
Hey @intoxicated 👋

Yes, you are right; this is the current way to do it.

**Comment by gmourier** [comment]
@intoxicated I'm moving your messages from https://github.com/meilisearch/product/discussions/568 in this discussion.

---

"I'm trying to migrate from Algolia to meilisearch, but I have been struggling to find a way to add unique element to array type.
For instance, let's say I have a document following:

{
   "id": 1234,
   "colors": ["red", "blue"]
}
and later I want to add some element to colors field but I wouldn't know which colors are already presented in the property unless I get the document first.

Algolia provides a way to add unique element to array field with batch like

{
  "op": "update",
  "partialUpdate": "addUnique",
  "id": 1234,
  "field_name": "colors",
  "value": "red"
}
so this document will attempt to update existing document's colors field with red. Since, red is already present, it won't do anything. Does meilisearch provide somewhat similar feature? if not, what's best way to update array instead update whole property"

"or maybe like https://discuss.elastic.co/t/how-to-update-a-object-within-array-in-elastic-search/20115/8"

---

**Comment by nmtri-dev** [comment]
Hey, I got a similar use case as mentioned in [discussion#3458](https://github.com/meilisearch/meilisearch/discussions/3458)

I have a scraper to scrape product details. The fields to update are dynamic and can be in nested object. Currently I have to fetch the whole product document, deep merge it with scraped data, then send the document back to Meilisearch.
It's good If Meilisearch can support append/ update nested object when updating document.

e.g. if I have a document like this
```json
{
    "price": {
       "current": 3000,
       "currency": "USD"
   }
}
```

And I do the update with this payload
```json
{
    "price": {
       "current": 4000
   }
}
```
The document should be updated like this
```json
{
    "price": {
       "current": 4000,
       "currency": "USD"
   }
}
```

Or I do the update with this payload
```json
{
    "price": {
       "sale": 2000
   }
}
```
The document should be updated like this
```json
{
    "price": {
       "current": 3000,
       "currency": "USD",
       "sale": 2000
   }
}
```
---
I object to support the same feature for removing the field in object, appending/ removing item in array. These actions are very hard to describe/ understand in simple data json payload, they should be done in application before sending to Meilisearch.

e.g. if you have a document like this
```json
{
    "tmpArray": [1, 2 ,3]
}
```

And you update with this payload
```json
{
   "tmpArray": ["a", "b", "c"]
}
```
Are you expecting the engine to replace the value or append the value?

To support these conveniences, I assume that the payload of update document api will have breaking changes.

**Comment by ShivamJoker** [comment]
I am also looking for the same feature.

**Comment by RemcoE33** [comment]
Looks like this is still a limitation. I my case I need to add localization field to a nested object. From the source those are 1 + <count locales>. In dot notation:

`freefields.namespace.color`
`freefields.namespace.color_FR`
`freefields.namespace.color_DE`
Etc..

I rather not combine this in the pipeline and keep those parts independent.

**Reply by curquiza** [reply]
Hey @RemcoE33 can you open an issue in https://github.com/meilisearch/meilisearch/issues with 
- what you do?
- what you expect?

Thanks
